tests/zfs-tests/tests/functional/hkdf/hkdf_test.c

```diff
@@ -0,0 +1,234 @@
+		.okm_len = 42,
+		.okm =	"\xce\x6c\x97\x19\x28\x05\xb3\x46"
+			"\xe6\x16\x1e\x82\x1e\xd1\x65\x67"
+			"\x3b\x84\xf4\x00\xa2\xb5\x14\xb2"
+			"\xfe\x23\xd8\x4c\xd1\x89\xdd\xf1"
+			"\xb6\x95\xb4\x8c\xbd\x1c\x83\x88"
+			"\x44\x11\x37\xb3\xce\x28\xf1\x6a"
+			"\xa6\x4b\xa3\x3b\xa4\x66\xb2\x4d"
+			"\xf6\xcf\xcb\x02\x1e\xcf\xf2\x35"
+			"\xf6\xa2\x05\x6c\xe3\xaf\x1d\xe4"
+			"\x4d\x57\x20\x97\xa8\x50\x5d\x9e"
+			"\x7a\x93",
+	},
+	{
+		.ikm =	"\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
+			"\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
+			"\x0b\x0b\x0b\x0b\x0b\x0b",
+		.ikm_len = 22,
+		.salt =	NULL,
+		.salt_len = 0,
+		.info =	NULL,
+		.info_len = 0,
+		.okm_len = 42,
+		.okm =	"\xf5\xfa\x02\xb1\x82\x98\xa7\x2a"
+			"\x8c\x23\x89\x8a\x87\x03\x47\x2c"
+			"\x6e\xb1\x79\xdc\x20\x4c\x03\x42"
+			"\x5c\x97\x0e\x3b\x16\x4b\xf9\x0f"
+			"\xff\x22\xd0\x48\x36\xd0\xe2\x34"
+			"\x3b\xac",
+	},
+	{
+		.ikm =	"\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
+			"\x0b\x0b\x0b",
+		.ikm_len = 11,
+		.salt =	"\x00\x01\x02\x03\x04\x05\x06\x07"
+			"\x08\x09\x0a\x0b\x0c",
+		.salt_len = 13,
+		.info =	"\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7"
+			"\xf8\xf9",
+		.info_len = 10,
+		.okm_len = 42,
+		.okm =	"\x74\x13\xe8\x99\x7e\x02\x06\x10"
+			"\xfb\xf6\x82\x3f\x2c\xe1\x4b\xff"
+			"\x01\x87\x5d\xb1\xca\x55\xf6\x8c"
+			"\xfc\xf3\x95\x4d\xc8\xaf\xf5\x35"
+			"\x59\xbd\x5e\x30\x28\xb0\x80\xf7"
+			"\xc0\x68",
+	},
+	{
+		.ikm =	"\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c"
+			"\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c"
+			"\x0c\x0c\x0c\x0c\x0c\x0c",
+		.ikm_len = 22,
+		.salt =	NULL,
+		.salt_len = 0,
+		.info = NULL,
+		.info_len = 0,
+		.okm_len = 42,
+		.okm =	"\x14\x07\xd4\x60\x13\xd9\x8b\xc6"
+			"\xde\xce\xfc\xfe\xe5\x5f\x0f\x90"
+			"\xb0\xc7\xf6\x3d\x68\xeb\x1a\x80"
+			"\xea\xf0\x7e\x95\x3c\xfc\x0a\x3a"
+			"\x52\x40\xa1\x55\xd6\xe4\xda\xa9"
+			"\x65\xbb",
+	},
+};
+
+static void
+hexdump(char *str, uint8_t *src, uint_t len)
+{
+	int i;
+
+	printf("\t%s\t", str);
+	for (i = 0; i < len; i++) {
+		printf("%02x", src[i] & 0xff);
+	}
+	printf("\n");
+}
+
+static int
+run_test(int i, hkdf_tv_t *tv)
+{
+	int ret;
+	uint8_t okey[SHA512_DIGEST_LENGTH];
+
+	printf("TEST %d:\t", i);
+
+	ret = hkdf_sha512((uint8_t *)tv->ikm, tv->ikm_len, (uint8_t *)tv->salt,
+	    tv->salt_len, (uint8_t *)tv->info, tv->info_len, okey, tv->okm_len);
+	if (ret != 0) {
+		printf("HKDF failed with error code %d\n", ret);
+		return (ret);
+	}
+
+	if (bcmp(okey, tv->okm, tv->okm_len) != 0) {
+		printf("Output Mismatch\n");
+		hexdump("Expected:", (uint8_t *)tv->okm, tv->okm_len);
+		hexdump("Actual:  ", okey, tv->okm_len);
+		return (1);
+	}
+
+	printf("Passed\n");
+
+	return (0);
+}
+
+int
+main(int argc, char **argv)
+{
+	int ret, i;
+
+	icp_init();
+
+	for (i = 0; i < NELEMS(test_vectors); i++) {
+		ret = run_test(i, &test_vectors[i]);
+		if (ret != 0)
+			goto error;
+	}
+
+	icp_fini();
+
+	printf("All tests passed successfully.\n");
+	return (0);
+
+error:
+	icp_fini();
+
+	printf("Test failed.\n");
+	return (ret);
+}
```

Comment:
Instead how about this.  Note that we always want to make sure we return `1` on failure.  Other non-zero values could result in problems when scripting if those values aren't well defined.

```c
int
main(int argc, char **argv)
{
	int error = 0;

        icp_init();

	for (int i = 0; i < NELEMS(test_vectors); i++) {
		error = run_test(i, &test_vectors[i]);
		if (error != 0)
			break;
	}

	icp_fini();

        if (error == 0) {
            	printf("All tests passed successfully.\n");
                return (0);
        } else {
        	printf("Test failed.\n");
                return (1);
        }
}
```